feat(masthead): overflow scrolling animation

ariellalgilmore opened the pull request:
### Related Ticket(s)

#4164 

### Description

create scrolling animation for overflow
![ezgif com-gif-maker](https://user-images.githubusercontent.com/20210594/97192043-c2f5f700-1764-11eb-8dfd-c77e290418d7.gif)

### Changelog

**New**

- additional styles
- use intersectionObserver to check for turning on and off left and right carets 

<!-- React and Web Component deploy previews are enabled by default. -->
<!-- To enable additional available deploy previews, apply the following -->
<!-- labels for the corresponding package: -->
<!-- *** "package: services": Services -->
<!-- *** "package: utilities": Utilities -->
<!-- *** "package: styles": Carbon Expressive -->
<!-- *** "RTL": React / Web Components (RTL) -->
<!-- *** "feature flag": React / Web Components (experimental) -->

asudoh commented on packages/react/src/components/Masthead/HeaderNavContainer.js:
```diff
@@ -18,97 +17,80 @@ const { prefix } = settings;
+        records => {
+          records.forEach(record => {
+            if (record.target.classList.contains('sub-content-left')) {
+              document.querySelector('.bx--header__nav-caret-left').hidden =
```
Should be grabbed from React `ref`.

asudoh commented on packages/react/src/components/Masthead/HeaderNavContainer.js:
```diff
@@ -18,97 +17,80 @@ const { prefix } = settings;
+                record.isIntersecting;
+            }
+            if (record.target.classList.contains('sub-content-right')) {
+              document.querySelector('.bx--header__nav-caret-right').hidden =
```
Should be grabbed from React `ref`.

asudoh commented on packages/react/src/components/Masthead/HeaderNavContainer.js:
```diff
@@ -18,97 +17,80 @@ const { prefix } = settings;
-        setShowLeftCaret(true);
-      }
+    if (io) {
+      io.observe(document.querySelector('.sub-content-left'));
+      io.observe(document.querySelector('.sub-content-right'));
```
Should be grabbed from React `ref`.

asudoh commented on packages/styles/scss/components/masthead/_masthead.scss:
```diff
@@ -307,6 +314,27 @@ $search-transition-timing: 95ms;
+    right: 1px;
+    width: 1px;
+    height: 100%;
+    //background-color: purple;
+    //clip: rect(0,0,0,0);
```
Did you mean to remove these lines?

asudoh commented on packages/styles/scss/components/masthead/_masthead.scss:
```diff
@@ -307,6 +314,27 @@ $search-transition-timing: 95ms;
+    left: 0;
+    width: 1px;
+    height: 100%;
+    //background-color: purple;
+
+    //clip: rect(0,0,0,0);
```
Did you mean to remove these lines?

asudoh reviewed (commented):
Great to see this seems ready-to-review @ariellalgilmore! 🎉 BTW did you mean to make this PR out of draft?

ariellalgilmore commented:
Thanks @asudoh!! Updated with refs and deleted comments from css! I left it as a draft because one thing I still was trying to research is why sometimes the right click needs to be pressed twice. Please let me know if you have any insights on this. Thanks!

asudoh commented:
Looks that the button visibility has adverse effect on container/content intersection state:

![intersection-caret-0](https://user-images.githubusercontent.com/1259051/97121524-3f3b0b00-1762-11eb-9f82-7a31d3bab5a0.gif)
![intersection-caret-1](https://user-images.githubusercontent.com/1259051/97121747-e0769100-1763-11eb-8679-da41677c5dc6.gif)

asudoh commented on packages/styles/scss/components/masthead/_masthead.scss:
```diff
@@ -308,13 +316,30 @@ $search-transition-timing: 95ms;
     }
   }
 
+  .sub-content-left {
+    position: absolute;
+    top: 0;
+    left: 0;
+    width: 1px;
+    height: 100%;
+  }
+
+  .sub-content-right {
+    position: absolute;
+    top: 0;
+    right: 0;
+    width: 1px;
+    height: 100%;
+  }
```
Do we want to add `.bx--` prefix? If not, probably `.dds--` prefix?

ariellalgilmore replied:
yes agree! Good catch!

ariellalgilmore replied:
just updated! Thanks!

asudoh reviewed (approved):
LGTM 👍 just one thing - Thanks @ariellalgilmore! 🎉

asudoh commented on packages/styles/scss/components/masthead/_masthead.scss:
```diff
@@ -308,13 +316,30 @@ $search-transition-timing: 95ms;
+    height: 100%;
+  }
+
+  .#{$prefix}-sub-content-right {
```
```suggestion
  .#{$prefix}--sub-content-right {
```

asudoh commented on packages/styles/scss/components/masthead/_masthead.scss:
```diff
@@ -308,13 +316,30 @@ $search-transition-timing: 95ms;
     }
   }
 
+  .#{$prefix}-sub-content-left {
```
```suggestion
  .#{$prefix}--sub-content-left {
```

asudoh commented on packages/react/src/components/Masthead/HeaderNavContainer.js:
```diff
@@ -18,100 +17,96 @@ const { prefix } = settings;
-        {children}
+      <div className={`${prefix}--header__nav-container`} ref={containerRef}>
+        <div className={`${prefix}--header__nav-content`} ref={contentRef}>
+          <div className={`${prefix}-sub-content-left`} ref={contentLeftRef} />
```
```suggestion
          <div className={`${prefix}--sub-content-left`} ref={contentLeftRef} />
```

asudoh commented on packages/react/src/components/Masthead/HeaderNavContainer.js:
```diff
@@ -18,100 +17,96 @@ const { prefix } = settings;
+        <div className={`${prefix}--header__nav-content`} ref={contentRef}>
+          <div className={`${prefix}-sub-content-left`} ref={contentLeftRef} />
+          <div
+            className={`${prefix}-sub-content-right`}
```
```suggestion
            className={`${prefix}--sub-content-right`}
```

asudoh reviewed (commented):
To follow the BEM naming structure - Thanks @ariellalgilmore!

jeffchew reviewed (changes requested):
Hey @ariellalgilmore , not sure if this is fully working, I can only click the arrow once and looks like there's still more items not fully visible.

![overflow](https://user-images.githubusercontent.com/1641214/97215674-24d15380-179b-11eb-8bca-c40a4d3ab036.gif)

ariellalgilmore commented:
closing due to conflicts. Will re-open new PR
